cmd/object-handlers.go

```diff
@@ -298,50 +285,57 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 
 	/// maximum Upload size for object in a single CopyObject operation.
 	if isMaxObjectSize(objInfo.Size) {
-		writeErrorResponse(w, r, ErrEntityTooLarge, objectSource)
-		return
-	}
-
-	// Size of object.
-	size := objInfo.Size
-
-	pipeReader, pipeWriter := io.Pipe()
-	go func() {
-		startOffset := int64(0) // Read the whole file.
-		// Get the object.
-		gErr := objectAPI.GetObject(sourceBucket, sourceObject, startOffset, size, pipeWriter)
-		if gErr != nil {
-			errorIf(gErr, "Unable to read an object.")
-			pipeWriter.CloseWithError(gErr)
+		writeErrorResponse(w, r, ErrEntityTooLarge, cpSrcPath)
+		return
+	}
+
+	cpMetadataOnly := strings.EqualFold(cpSrcPath, cpDestPath)
+	var newMetadata map[string]string
+	// Source and destination objects cannot be same if x-amz-metadata-directive is not set.
+	if cpMetadataOnly {
+		// Check if x-amz-metadata-direction was set to replace metadata.
+		if isMetadataReplace(r.Header) {
+			// Replace metadata requested proceed to extract
+			// relevant metadata from the request headers.
+			newMetadata = extractMetadataFromHeader(r.Header)
+
+			// Check if x-amz-metadata-direction was set to copy metadata.
+		} else if isMetadataCopy(r.Header) {
```

Comment:
this is an error case as well:

`You cannot copy an object to itself unless the MetadataDirective header is specified and its value set to REPLACE.`